Platform/OpenCanopy/BitmapFont.c

```diff
@@ -398,68 +413,72 @@ BmfContextInitialize (
     return FALSE;
   }
 
-  Context->Height  = (UINT16) Height;
+  Context->Height  = Context->Common->lineHeight;
   Context->OffsetY = -MinY;
 
   Pairs = Context->KerningPairs;
-  for (Index = 0; Index < Context->NumKerningPairs; ++Index) {
-    Char = BmfGetChar (Context, Pairs[Index].first);
-    if (Char == NULL) {
-      DEBUG ((
-        DEBUG_WARN,
-        "BMF: Pair char %u not found\n",
-        Pairs[Index].first
-        ));
-      return FALSE;
-    }
-
-    Result = OcOverflowAddS32 (
-               Char->xoffset + Char->width,
-               Pairs[Index].amount,
-               &Width
-               );
-    Result |= OcOverflowAddS32 (
-                Char->xoffset + Char->xadvance,
-                Pairs[Index].amount,
-                &Advance
-                );
-    if (Result
-     || 0 > Width || Width > MAX_UINT16
-     || 0 > Advance || Advance > MAX_UINT16) {
-       DEBUG ((
-         DEBUG_WARN,
-         "BMF: Pair insane\n"
-         " first %u\n"
-         " second %u\n"
-         " amount %d\n",
-         Pairs[Index].first,
-         Pairs[Index].second,
-         Pairs[Index].amount
-         ));
-      return FALSE;
-    }
-    //
-    // This only yields unexpected but not undefined behaviour when not met,
-    // hence it is fine verifying it only DEBUG mode.
-    //
-    DEBUG_CODE_BEGIN ();
-    if (Index > 0) {
-      if (Pairs[Index - 1].first > Pairs[Index].first) {
-        DEBUG ((DEBUG_WARN, "BMF: First Character IDs are not sorted\n"));
+  if (Pairs != NULL) { // According to the docs, kerning pairs are optional
+    for (Index = 0; Index < Context->NumKerningPairs; ++Index) {
+      Char = BmfGetChar (Context, Pairs[Index].first);
+      if (Char == NULL) {
+        DEBUG ((
+          DEBUG_WARN,
+          "BMF: Pair char %u not found\n",
+          Pairs[Index].first
+          ));
         return FALSE;
       }
 
-      if (Pairs[Index - 1].first == Pairs[Index].first) {
-        if (Pairs[Index - 1].second > Pairs[Index].second) {
-          DEBUG ((DEBUG_WARN, "BMF: Second Character IDs are not sorted\n"));
-          return FALSE;
-        } else if (Pairs[Index - 1].second == Pairs[Index].second) {
-          DEBUG ((DEBUG_WARN, "BMF: Second Character ID duplicate\n"));
+      Result = OcOverflowAddS32 (
+                 Char->xoffset + Char->width,
+                 Pairs[Index].amount,
+                 &Width
+                 );
+      Result |= OcOverflowAddS32 (
+                  Char->xoffset + Char->xadvance,
+                  Pairs[Index].amount,
+                  &Advance
+                  );
+      if (Result
+       || 0 > Width || Width > MAX_UINT16
+       || 0 > Advance || Advance > MAX_UINT16) {
+         DEBUG ((
+           DEBUG_WARN,
+           "BMF: Pair at index %d insane\n"
+           " first %u\n"
```

Comment:
Please remove newlines here actually as we do not support multiline logging. Just spaces will be fine.

Reply:
Not my change again (I just added index) but I'll fix.

Reply:
Thanks!